cps_data/pycps/impute.py

```diff
@@ -0,0 +1,262 @@
+import pandas as pd
+import numpy as np
+from scipy.stats import norm
+from helpers import log, LOG_VARS, X_VARS, FILINGPARAMS, CPS_YR_IDX
+
+
+def impute(df, logit_betas, ols_betas, x_vars, l_adj, o_adj, prob_mult):
+    """
+    This function first determines if a given record would have claimed a
+    deduction, then determines the value of that deduction for each record
+
+    Parameters
+    ----------
+    df: dataframe will all of the records
+    logit_betas: beta coefficients from the logit model
+    ols_betas: beta coefficients from the ols model
+    x_vars: x variables used in the imputation
+    l_adj: specific adjustment added when determining logit probabilities
+    o_adj: specific adjustment added during ols calculations
+    prob_mult: multiplier for the probabilities from the logit model
+
+    Returns
+    -------
+    Array of imputed values
+    """
+    # calculate probability using coefficients from the logit model
+    xb = np.zeros(len(df))
+    for param in x_vars:
+        xb += df[param] * logit_betas[param]
+    prob = np.exp(xb) / (1. + np.exp(xb)) + l_adj
+    prob *= prob_mult
+
+    # flag which records should recieve and imputation
+    z1 = np.random.uniform(0, 1, len(prob))
+    df["impute"] = np.where(z1 <= prob, 1, 0)
+    impute = np.where(z1 <= prob, 1, 0)
+    z2 = np.random.randn(len(prob))
+    xb = np.zeros(len(df))
+    for param in x_vars:
+        xb += df[param] * ols_betas[param]
+    val = np.exp(xb + o_adj * z2) * impute
+    return val
+
+
+def tobit(cps, betas, x_vars, sigma, prob_mult):
+    """
+    Impute values using a tobit model
+
+    Parameters
+    ----------
+    cps: DataFrame with all CPS records
+    betas: beta coefficients from the tobit regression
+    x_vars: x variables used in the imputations
+    sigma: value used in tobit calculations
+    prob_mult: multiplier for probability values
+
+    Returns
+    -------
+    val: an array of values for the imputed variable
+    """
+    xb = np.zeros(len(cps))
+    for param in x_vars:
+        xb += cps[param] * betas[param]
+
+    prob = norm.cdf(xb / sigma) * prob_mult
+    z1 = np.random.uniform(0, 1, len(prob))
+    impute = np.where(z1 <= prob, 1, 0)
+    lamb = norm.pdf(xb / sigma) / norm.cdf(xb / sigma)
+    val = (xb + sigma * lamb) * impute
+    return val
+
+
+def imputation(data, logit_betas, ols_betas):
+    """
+    This function uses beta values calculated using the IRS Public Use File
+    to impute the value of certain itemized deductions for filing records
+    """
+    # data prep
+    # find log of specified variables
+    for var in LOG_VARS:
+        # use lowercase because CPS uses lowercase at this stage
+        data[f"ln{var.lower()}"] = log(data, var.lower())
+
+    data["joint_filer"] = np.where(
+        data["mars"] == 2, 1, 0
+    )
+    # cap family size at 5 to match with the PUF
+    # data["fam_size"] = np.minimum(data["XTOT"], 5)
```

Comment:
is family size being capped elsewhere? if not could you uncomment the code, and possibly leave a comment re: why it's not being done?

Reply:
Good catch! Should be uncommented.